File: src/com/github/doomkin/jsubsetsum/CombinationsSequence.java
/**
 * Class CombinationsSequence enumerates k-combinations.
 * 
 * @author Pavel Nikitin
 * License: MIT
 */
package com.github.doomkin.jsubsetsum;

import java.util.AbstractCollection;
import java.util.Iterator;

public class CombinationsSequence extends AbstractCollection<Combination> 
{
    private int _setSize;
    private int _subsetSize;

    public CombinationsSequence(int setSize, int subsetSize)
    {
        _setSize = setSize;
        _subsetSize = subsetSize;
    }

    @Override
    public Iterator<Combination> iterator() 
    {
        return new Iterator<Combination>() 
        {
            private Combination _combination;
            private int _index = 0;
            private boolean _computed = false;
            
            {
                // create initial combination
                _combination = new Combination(_subsetSize);
                
                // reduce the last index for first iteration of computeNext()
                _combination.set(_subsetSize - 1, _subsetSize - 2);
            }
            
            public boolean hasNext() 
            {
                if (!_computed)
                    computeNext();
      
                return _index < _subsetSize;
            }
            
            public Combination next() 
            {
                if (hasNext())
                    _computed = false;
                else
                    throw new ArrayIndexOutOfBoundsException(
                            "End of a sequence of combinations.");
                
                return _combination;
            }
            
            public void remove() 
            { 
                throw new UnsupportedOperationException(
                        "Cannot remove combination from sequence."); 
            }
            
            // Compute next combination
            private void computeNext()
            {
                // find the first item that has not reached its maximum value
                _index = _subsetSize;
                
                for (int i = _subsetSize - 1; i >= 0; i--)
                    if (_combination.get(i) < _setSize - _subsetSize + i)
                    {
                        _index = i;
                        break;
                    }
                
                // there is an item which has not reached its maximum value
                if (_index < _subsetSize)
                {    
                    // generate the next combination in lexographical order
                    _combination.set(_index, _combination.get(_index) + 1);
    
                    for (int i = _index + 1; i < _subsetSize; i++)
                        _combination.set(i, _combination.get(i-1) + 1);
                }
                
                _computed = true;
            }
        };
    }

    // Size of index of the binomial coefficient (in bits)
    @Override
    public int size() 
    {
        return lg2(bc());
    }
    
    // Compute the logarithm of number n to the base 2
    public static int lg2(long n)
    {
        int result;
        for (result = 0; n > 0; n >>= 1, result++);
        return result; 
    }

    // Compute the binomial coefficient indexed by setSize and subsetSize
    public long bc()
    {
        return bc(_setSize, _subsetSize);
    }

    // Compute the binomial coefficient indexed by n and k
    private static long bc(int n, int k)
    {
        return k == 0 ? 1 : (n * bc(n - 1, k - 1)) / k;
    }
}

File: src/CombinationsSequence.kt
/** Перебор k-комбинаций */

import java.util.AbstractCollection

class CombinationsSequence(private val setSize: Int, private val subsetSize: Int) :
  AbstractCollection<Combination>() {

  override fun iterator(): MutableIterator<Combination> {
    return object : MutableIterator<Combination> {
      private var combination: Combination = Combination(subsetSize)
      private var index = 0
      private var computed = false

      init {
        combination[subsetSize - 1] = subsetSize - 2
      }

      override fun hasNext(): Boolean {
        if (!computed)
          computeNext()

        return index < subsetSize
      }

      override fun next(): Combination {
        if (hasNext())
          computed = false
        else
          throw ArrayIndexOutOfBoundsException("Конец перебора комбинаций")

        return combination
      }

      override fun remove() {
        throw UnsupportedOperationException("Ошибка удаления комбинации из последовательности комбинаций")
      }

      private fun computeNext() {
        // Поиск первого элемента, не достигшего своего максимального значения
        index = subsetSize

        for (i in subsetSize - 1 downTo 0)
          if (combination[i] < setSize - subsetSize + i) {
            index = i
            break
          }

        // Найден элемент, не достигший своего максимального значения
        if (index < subsetSize) {
          // Поиск следующей комбинации в лексикографическом порядке
          combination[index] = combination[index] + 1

          for (i in index + 1 until subsetSize)
            combination[i] = combination[i - 1] + 1
        }

        computed = true
      }
    }
  }

  // Размер биномиального коэффициента в битах
  override val size: Int
    get() = log2(binomialCoefficient(setSize, subsetSize))

  // Число сочетаний из n по k
  private fun binomialCoefficient(n: Int, k: Int): Long =
    if (k == 0) 1 else n * binomialCoefficient(n - 1, k - 1) / k

  companion object {
    @JvmStatic
    fun log2(n: Long): Int {
      var m = n
      var result = 0
      while (m > 0) {
        m = m shr 1
        result++
      }
      return result
    }
  }
}